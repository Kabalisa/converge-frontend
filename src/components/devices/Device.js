import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';

const formatDate = date => moment(date).format('DD MMM YYYY');

/**
 *
 * Device component
 *
 * @param {Object} device
 *
 * @returns {JSX}
 */
const Device = ({
  device: {
    name, deviceType, dateAdded, lastSeen, location, /* room: { name: roomName }, */
  },
}) => (
  <div className="table__row">
    <span>{name}</span>
    <span>{deviceType}</span>
    <span>{formatDate(dateAdded)}</span>
    <span>{formatDate(lastSeen)}</span>
    <span>{location}</span>
    {/* <span>{roomName}</span> */}
  </div>
);

Device.propTypes = {
  device: PropTypes.shape({
    name: PropTypes.string.isRequired,
    deviceType: PropTypes.string.isRequired,
    dateAdded: PropTypes.string.isRequired,
    lastSeen: PropTypes.string.isRequired,
    location: PropTypes.string.isRequired,
  }).isRequired,
};

export default Device;
