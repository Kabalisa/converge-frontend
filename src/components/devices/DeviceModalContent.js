import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Input, SelectInput, ActionButtons } from '../commons';

const DeviceModalContent = ({
  formData: {
    name, deviceType, roomId,
  },
  methods: {
    handleNameChange, handleTypeChange, handleRoomChange, closeModal, submit,
  },
  modalContent, rooms, deviceTypes, fetching,
}) => (
  <form className="amenity-form device-type">
    <div>
      {
        modalContent !== 'delete'
          ? (
            <Fragment>
              <Input
                id="device-name"
                name="device-name"
                placeholder="Enter Device Name"
                labelName="Device Name"
                labelClass="device-input"
                value={name}
                onChange={handleNameChange}
              />
              <SelectInput
                labelText="Device Type"
                name="device-type"
                id="add-devices-type"
                isValue
                value={deviceType || 'Select Device'}
                options={deviceTypes}
                onChange={handleTypeChange}
                wrapperClassName="device-input"
                placeholder="Select Device"
                selectInputClassName="dynamic-input-field default-select"
              />
              <SelectInput
                labelText="Room"
                name="room"
                id="add-devices-rooms"
                value={roomId || 'Select Room'}
                options={rooms}
                onChange={handleRoomChange}
                wrapperClassName="device-input"
                placeholder="Select Room"
                selectInputClassName="dynamic-input-field default-select"
              />
            </Fragment>
          ) : (
            <p id="confirm-msg">
              Are you sure you want to delete &quot;{`${name}: ${deviceType}`}
              &quot;? <br />
              This cannot be undone
            </p>
          )
      }
      <div className="loading-btn-div">
        <ActionButtons
          withCancel
          isLoading={fetching}
          onClickCancel={closeModal}
          actionButtonText={modalContent && ((modalContent === 'edit' ? 'SAVE CHANGES' : `${modalContent.toUpperCase()} DEVICE`) || 'ADD DEVICE')}
          onClickSubmit={submit}
        />
      </div>
    </div>
  </form>
);

DeviceModalContent.propTypes = {
  formData: PropTypes.shape({
    name: PropTypes.string,
    roomId: PropTypes.string,
    deviceType: PropTypes.string,
  }).isRequired,
  methods: PropTypes.shape({
    handleNameChange: PropTypes.func,
    handleRoomChange: PropTypes.func,
    handleTypeChange: PropTypes.func,
    closeModal: PropTypes.func,
    submit: PropTypes.func,
  }).isRequired,
  fetching: PropTypes.bool,
  modalContent: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  rooms: PropTypes.instanceOf(Array).isRequired,
  deviceTypes: PropTypes.instanceOf(Array).isRequired,
};

DeviceModalContent.defaultProps = {
  fetching: false,
  modalContent: undefined,
};

export default DeviceModalContent;
