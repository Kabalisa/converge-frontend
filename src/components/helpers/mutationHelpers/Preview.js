import apolloClient from '../../../utils/ApolloClient';
import { ADD_LEVEL_SETUP_MUTATION } from '../../../graphql/mutations/Preview';
import GET_ALL_LEVELS from '../../../graphql/queries/Levels';

const addLevelSetup = async ({ flattenedData }, client = apolloClient) => {
  await client.mutate({
    mutation: ADD_LEVEL_SETUP_MUTATION,
    name: 'preview',
    variables: {
      flattenedData,
    },
    update: async (store) => {
      // Read the data from our cache for this query.
      const data = store.readQuery({ query: GET_ALL_LEVELS });
      data.allStructures = [...data.allStructures, flattenedData];
      store.writeQuery({
        query: GET_ALL_LEVELS,
        data,
      });
    },
  });
};

export default addLevelSetup;
