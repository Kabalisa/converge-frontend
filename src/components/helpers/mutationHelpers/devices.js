/* eslint-disable import/prefer-default-export */
import apolloClient from '../../../utils/ApolloClient';
import { ADD_DEVICE_MUTATION } from '../../../graphql/mutations/Devices';
import { GET_DEVICES_QUERY } from '../../../graphql/queries/Devices';

export const addDeviceMutation = async ({
  name, roomId, deviceType, location,
}, client = apolloClient) => {
  await client
    .mutate({
      mutation: ADD_DEVICE_MUTATION,
      name: 'addDeviceMutation',
      variables: {
        name, deviceType, roomId, location,
      },
      // write and read from cache
      update: async (proxy, { data: { createDevice } }) => {
        const cachedDevices = proxy.readQuery({
          query: GET_DEVICES_QUERY,
        });
        const deviceList = cachedDevices.allDevices;
        cachedDevices.allDevices = [...deviceList, createDevice.device];

        proxy.writeQuery({
          query: GET_DEVICES_QUERY,
          data: cachedDevices,
        });
      },
    });
};
