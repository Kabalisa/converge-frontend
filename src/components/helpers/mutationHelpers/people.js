/* eslint-disable import/prefer-default-export */
import apolloClient from '../../../utils/ApolloClient';
import { INVITE_PERSON_MUTATION, CHANGE_USER_LOCATION_MUTATION } from '../../../graphql/mutations/People';
import { decodeTokenAndGetUserData } from '../../../utils/Cookie';
import GET_ALL_LEVELS from '../../../graphql/queries/Levels';


const invitePersonMutation = async (email, client = apolloClient) => {
  await client
    .mutate({
      mutation: INVITE_PERSON_MUTATION,
      name: 'invitePersonMutation',
      variables: { email },
    });
};

const changeUserLocation = async (locationId, queryEmail, client = apolloClient) => {
  const { UserInfo: { email: tokenEmail } } = decodeTokenAndGetUserData();
  const email = queryEmail || tokenEmail;

  const { data } = await client
    .mutate({
      mutation: CHANGE_USER_LOCATION_MUTATION,
      name: 'changeUserLocation',
      variables: { email, locationId },
      refetchQueries: [{
        query: GET_ALL_LEVELS,
      },
      ],
    });

  return data;
};

export { invitePersonMutation, changeUserLocation };
