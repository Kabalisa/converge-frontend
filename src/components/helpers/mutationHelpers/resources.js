/* eslint-disable import/prefer-default-export */
/* eslint-disable consistent-return */
/* eslint-disable array-callback-return */
import apolloClient from '../../../utils/ApolloClient';
import { DELETE_RESOURCE_MUTATION } from '../../../graphql/mutations/resources';
import { GET_RESOURCES_QUERY } from '../../../graphql/queries/Resources';

const updateStore = (data, deleteResource) =>
  data.allResources.resources.filter((resource) => {
    if (resource.id !== deleteResource.resource.id) {
      return resource;
    }
  });

const deleteResources = async (currentPage, perPage, { resourceId }, client = apolloClient) => {
  await client
    .mutate({
      mutation: DELETE_RESOURCE_MUTATION,
      name: 'deleteResource',
      variables: {
        resourceId,
      },
      refetchQueries: [
        {
          query: GET_RESOURCES_QUERY,
          variables: {
            page: currentPage,
            perPage,
          },
        },
      ],
      update: async (proxy, { data: { deleteResource } }) => {
        // Read the data from our cache for this query.
        const resourceData = proxy.readQuery({
          query: GET_RESOURCES_QUERY,
          variables: {
            page: currentPage,
            perPage,
          },
        });
        // Remove a resource from the store.
        resourceData.allResources.resources = updateStore(resourceData, deleteResource);
        // Write our data back to the cache.
        proxy.writeQuery({
          query: GET_RESOURCES_QUERY,
          variables: {
            page: currentPage,
            perPage,
          },
          data: resourceData,
        });
      },
    });
};
export { deleteResources, updateStore };
