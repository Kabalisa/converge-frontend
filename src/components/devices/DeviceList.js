import React from 'react';
import '../../assets/styles/devicelist.scss';
import ColGroup from '../helpers/ColGroup';
import TableHead from '../helpers/TableHead';
import Device from './Device';
import devices from '../../fixtures/devices';
import MenuTitle from '../commons/MenuTitle';

const deviceListItems = devices.map(device => (
  <Device device={device} key={device.name} />
));

/**
 * Device List Component
 *
 * @returns {JSX}
 */
const DeviceList = () => (
  <div className="settings-devices-list">
    <div className="settings-devices-control">
      <MenuTitle title="Devices" />
    </div>
    <table>
      <ColGroup />
      <TableHead
        titles={['Name', 'Type', 'Date Added', 'Last Seen', 'Location']}
      />
      <tbody>{deviceListItems}</tbody>
    </table>
  </div>
);

export default DeviceList;
