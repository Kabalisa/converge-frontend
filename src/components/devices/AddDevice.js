import React, { Component } from 'react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import notification from '../../utils/notification';
import MrmModal from '../commons/Modal';
import { ActionButtons, Input, SelectInput } from '../commons';
import { addDeviceMutation } from '../helpers/mutationHelpers/devices';

class DeviceFormModal extends Component {
  state = {
    isModalClosed: false,
    name: '',
    roomId: '',
    deviceType: '',
  }

  componentDidMount() {
    const { device } = this.props;
    if (device) this.updateStateWithPropDevice();
  }

  getDeviceTypes = () => ([{
    name: '7"',
  }, {
    name: '8"',
  }, {
    name: '9.7"',
  }, {
    name: '10.1"',
  }])

  updateStateWithPropDevice = () => {
    const { name, roomId, deviceType } = this.props.device;
    this.setState({
      name, roomId, deviceType,
    });
  }

  handleNameChange = ({ target: { value: name } }) => {
    this.setState({ name });
  };

  handleTypeChange = ({ target: { value: deviceType } }) => {
    this.setState({ deviceType });
  };

  handleRoomChange = ({ target: { value: roomId } }) => {
    this.setState({ roomId });
  };

  closeModal = (e) => {
    if (e) {
      e.preventDefault();
    }
    this.setState({ isModalClosed: true });
  };

  handleModalCloseRequest = () => {
    this.state.isModalClosed && this.setState({ isModalClosed: false });
  };

  submit = async (e) => {
    e.preventDefault();
    const { name, roomId, deviceType } = this.state;
    const { location, refetch } = this.props;
    if (
      !name || !roomId || !deviceType
    ) {
      return notification(toastr, 'error', 'Please fill all fields')();
    }
    try {
      this.setState({ fetching: true });
      await addDeviceMutation({
        name, roomId, location: location.name, deviceType,
      });
      await refetch();
      this.setState({
        fetching: false,
        isModalClosed: true,
        name: '',
        deviceType: '',
      });

      return notification(toastr, 'success', 'Device successfully added.')();
    } catch (error) {
      notification(toastr, 'error', 'An error occured while adding device.')();
      return this.setState({ fetching: false });
    }
  };

  render() {
    const {
      isModalClosed, name, fetching,
    } = this.state;
    const { rooms } = this.props;
    return (
      <MrmModal
        title="ADD A NEW DEVICE"
        buttonText="Add a New Device"
        closeModal={isModalClosed}
        handleCloseRequest={this.handleModalCloseRequest}
        className="add-resource-modal"
      >
        <form className="amenity-form device-type">
          <div>
            <Input
              id="device-name"
              name="device-name"
              placeholder="Enter Device Name"
              labelName="Device Name"
              labelClass="device-input"
              value={name}
              onChange={this.handleNameChange}
            />
            <SelectInput
              labelText="Device Type"
              name="device-type"
              id="add-devices-type"
              isValue
              value="Select Device"
              options={this.getDeviceTypes()}
              onChange={this.handleTypeChange}
              wrapperClassName="device-input"
              placeholder="Device Type"
              selectInputClassName="dynamic-input-field default-select"
            />
            <SelectInput
              labelText="Room"
              name="room"
              id="add-devices-rooms"
              value="Select Room"
              options={rooms}
              onChange={this.handleRoomChange}
              wrapperClassName="device-input"
              placeholder="Select Room"
              selectInputClassName="dynamic-input-field default-select"
            />
            <div className="loading-btn-div">
              <ActionButtons
                withCancel
                isLoading={fetching}
                onClickCancel={this.closeModal}
                actionButtonText="ADD DEVICE"
                onClickSubmit={this.submit}
              />
            </div>
          </div>
        </form>
      </MrmModal>
    );
  }
}

DeviceFormModal.propTypes = {
  refetch: PropTypes.func.isRequired,
  location: PropTypes.shape({
    name: PropTypes.string,
  }),
  rooms: PropTypes.instanceOf(Array),
  device: PropTypes.shape({
    name: PropTypes.string,
    deviceType: PropTypes.string,
    roomId: PropTypes.number,
  }),
};

DeviceFormModal.defaultProps = {
  location: undefined,
  rooms: [],
  device: undefined,
};

export default DeviceFormModal;
