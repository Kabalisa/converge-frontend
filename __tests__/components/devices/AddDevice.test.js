import React from 'react';
import { shallow } from 'enzyme';
import toastr from 'toastr';
import DeviceFormModal from '../../../src/components/devices/AddDevice';
import { addDeviceMutation } from '../../../src/components/helpers/mutationHelpers/devices';


jest.mock('../../../src/components/helpers/mutationHelpers/devices');
jest.mock('toastr');

describe('Tests for DeviceFormModal', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<DeviceFormModal
      location={{
      name: 'Lagos',
    }}
      refetch={jest.fn()}
    />);
  });

  it('Should render correctly from memory', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should handle submit correctly with correct data', async () => {
    addDeviceMutation.mockResolvedValue({});
    wrapper.setState({
      name: 'Galaxy S9',
      roomId: '32',
      deviceType: 'Tablet',
    });
    await wrapper.instance().submit({ preventDefault: jest.fn() });
    expect(addDeviceMutation).toHaveBeenCalled();
    expect(toastr.success).toHaveBeenCalled();
  });

  it('Should handle server errors', async () => {
    wrapper.setState({
      name: 'Galaxy S9',
      roomId: '32',
      deviceType: 'Tablet',
    });
    addDeviceMutation.mockRejectedValue({});
    await wrapper.instance().submit({ preventDefault: jest.fn() });
    expect(toastr.error).toHaveBeenCalledWith('An error occured while adding device.');
  });


  it('Should not submit with incomplete data', () => {
    wrapper.setState({
      deviceType: undefined,
    });
    wrapper.instance().submit({ preventDefault: jest.fn() });
    expect(toastr.error).toHaveBeenCalledWith('Please fill all fields');
  });

  it('Should call setter methods', () => {
    const getInputObject = value => ({
      target: {
        value,
      },
    });

    wrapper.instance().handleNameChange(getInputObject('Tab'));
    expect(wrapper.state('name')).toBe('Tab');

    wrapper.instance().handleTypeChange(getInputObject('Tablet'));
    expect(wrapper.state('deviceType')).toBe('Tablet');

    wrapper.instance().handleRoomChange(getInputObject('123'));
    expect(wrapper.state('roomId')).toBe('123');
  });

  it('Should use device if provided', () => {
    wrapper.setProps({
      device: {
        name: 'Galaxy S10',
        roomId: '31',
        deviceType: 'Phone',
      },
    });
    wrapper.instance().componentDidMount();
    expect(wrapper.state('roomId')).toBe('31');
  });

  it('Should close modal', () => {
    wrapper.instance().closeModal();
    wrapper.instance().closeModal({ preventDefault: jest.fn() });
    expect(wrapper.state('isModalClosed')).toBe(true);

    wrapper.instance().handleModalCloseRequest();
    expect(wrapper.state('isModalClosed')).toBe(false);
  });
});
