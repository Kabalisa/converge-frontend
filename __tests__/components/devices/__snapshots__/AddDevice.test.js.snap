// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for DeviceFormModal Should render correctly from memory 1`] = `
<MrmModal
  buttonText="Add a New Device"
  className="add-resource-modal"
  closeModal={false}
  handleCloseRequest={[Function]}
  modalButtonClassName="button"
  title="ADD A NEW DEVICE"
>
  <form
    className="amenity-form device-type"
  >
    <div>
      <Input
        controlsClass=""
        id="device-name"
        inputClass="mrm-input default-input"
        labelClass="device-input"
        labelName="Device Name"
        name="device-name"
        onChange={[Function]}
        placeholder="Enter Device Name"
        type="text"
        value=""
      />
      <SelectInput
        id="add-devices-type"
        isValue={true}
        labelText="Device Type"
        name="device-type"
        onChange={[Function]}
        onClick={[Function]}
        options={
          Array [
            Object {
              "name": "7\\"",
            },
            Object {
              "name": "8\\"",
            },
            Object {
              "name": "9.7\\"",
            },
            Object {
              "name": "10.1\\"",
            },
          ]
        }
        placeholder="Device Type"
        placeholderValue=""
        required={false}
        selectInputClassName="dynamic-input-field default-select"
        value="Select Device"
        wrapperClassName="device-input"
      />
      <SelectInput
        id="add-devices-rooms"
        isValue={false}
        labelText="Room"
        name="room"
        onChange={[Function]}
        onClick={[Function]}
        options={Array []}
        placeholder="Select Room"
        placeholderValue=""
        required={false}
        selectInputClassName="dynamic-input-field default-select"
        value="Select Room"
        wrapperClassName="device-input"
      />
      <div
        className="loading-btn-div"
      >
        <ActionButtons
          actionButtonText="ADD DEVICE"
          cancelButtonText="CANCEL"
          isLoading={false}
          onClickCancel={[Function]}
          onClickSubmit={[Function]}
          showActionButton={true}
          withCancel={true}
        />
      </div>
    </div>
  </form>
</MrmModal>
`;
